lib/core/CProgramCounters.cc

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <core/CProgramCounters.h>
+
+#include <core/CLogger.h>
+#include <core/CRapidJsonLineWriter.h>
+#include <core/CStatePersistInserter.h>
+#include <core/CStateRestoreTraverser.h>
+#include <core/CStringUtils.h>
+
+#include <rapidjson/document.h>
+#include <rapidjson/ostreamwrapper.h>
+
+#include <ostream>
+#include <string>
+
+namespace ml {
+namespace core {
+
+namespace {
+
+using TGenericLineWriter = core::CRapidJsonLineWriter<rapidjson::OStreamWrapper>;
+
+const std::string NAME_TYPE("name");
+const std::string DESCRIPTION_TYPE("description");
+const std::string COUNTER_TYPE("value");
+
+//! Persistence tags
+const std::string KEY_TAG("a");
+const std::string VALUE_TAG("b");
+
+//! Helper function to add a string/int pair to JSON writer
+void addStringInt(TGenericLineWriter& writer,
+                  const std::string& name,
+                  const std::string& description,
+                  uint64_t counter) {
+    writer.StartObject();
+
+    writer.String(NAME_TYPE);
+    writer.String(name);
+
+    writer.String(DESCRIPTION_TYPE);
+    writer.String(description);
+
+    writer.String(COUNTER_TYPE);
+    writer.Uint64(counter);
+
+    writer.EndObject();
+}
+}
+
+CProgramCounters::CProgramCounters()
+    : m_Counters(), m_Cache(), m_ProgramCounterTypes() {
+}
+
+CProgramCounters& CProgramCounters::instance() {
+    return ms_Instance;
+}
+
+CProgramCounters::TCounter& CProgramCounters::counter(counter_t::ECounterTypes counterType) {
+    if (counterType >= counter_t::E_LastEnumCounter) {
+        LOG_ABORT(<< "Bad index " << counterType);
+    }
+    return ms_Instance.m_Counters[counterType];
+}
+
+CProgramCounters::TCounter& CProgramCounters::counter(size_t index) {
+    if (index >= ms_Instance.m_Counters.size()) {
+        LOG_ABORT(<< "Bad index " << index);
```

Comment:
This can't be completely ignored: this has to return a writable reference, but a catch all counter for this case should do the trick. Alternatively, this could be reworked as an increment function and then truly ignored, but that is a bigger change.

Reply:
Since the method returns `CProgramCounters::TCounter&` we'll have to return _something_ if we ignore out-of-bounds errors.  Similar to what @tveasey said in https://github.com/elastic/ml-cpp/pull/369/files#r252367258 I would suggest having a "bad counter" member of the class that's separate to the array of valid counters.  We can log a warning whenever this is accessed, but not abort the process.